% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  11pt,
]{article}

\usepackage{amsmath,amssymb}
\usepackage{setspace}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else  
    % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1.0in]{geometry}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
% Make \paragraph and \subparagraph free-standing
\makeatletter
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}{
    \@ifstar
      \xxxParagraphStar
      \xxxParagraphNoStar
  }
  \newcommand{\xxxParagraphStar}[1]{\oldparagraph*{#1}\mbox{}}
  \newcommand{\xxxParagraphNoStar}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}{
    \@ifstar
      \xxxSubParagraphStar
      \xxxSubParagraphNoStar
  }
  \newcommand{\xxxSubParagraphStar}[1]{\oldsubparagraph*{#1}\mbox{}}
  \newcommand{\xxxSubParagraphNoStar}[1]{\oldsubparagraph{#1}\mbox{}}
\fi
\makeatother


\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
% definitions for citeproc citations
\NewDocumentCommand\citeproctext{}{}
\NewDocumentCommand\citeproc{mm}{%
  \begingroup\def\citeproctext{#2}\cite{#1}\endgroup}
\makeatletter
 % allow citations to break across lines
 \let\@cite@ofmt\@firstofone
 % avoid brackets around text for \cite:
 \def\@biblabel#1{}
 \def\@cite#1#2{{#1\if@tempswa , #2\fi}}
\makeatother
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newenvironment{CSLReferences}[2] % #1 hanging-indent, #2 entry-spacing
 {\begin{list}{}{%
  \setlength{\itemindent}{0pt}
  \setlength{\leftmargin}{0pt}
  \setlength{\parsep}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1
   \setlength{\leftmargin}{\cslhangindent}
   \setlength{\itemindent}{-1\cslhangindent}
  \fi
  % set entry spacing
  \setlength{\itemsep}{#2\baselineskip}}}
 {\end{list}}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{\hfill\break\parbox[t]{\linewidth}{\strut\ignorespaces#1\strut}}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{\strut#1\strut}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{\strut#1\strut}}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

\usepackage[left]{lineno}
\linenumbers
\modulolinenumbers
\usepackage{helvet}
\renewcommand*\familydefault{\sfdefault}
\usepackage[T1]{fontenc}
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother

\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}

\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}


\author{}
\date{}

\begin{document}


\setstretch{1.75}
\raggedright

\section{phylotypr: An R package for classifying DNA
sequences}\label{phylotypr-an-r-package-for-classifying-dna-sequences}

\vspace{20mm}

\textbf{Running title:} phylotypr

\vspace{20mm}

Patrick D. Schloss\textsuperscript{\textdagger}

\vspace{25mm}

\textdagger To whom correspondence should be addressed\\
\href{mailto:pschloss@umich.edu}{pschloss@umich.edu}

\vspace{10mm}

Department of Microbiology \& Immunology\\
University of Michigan\\
Ann Arbor, MI 48109

\vspace{20mm}

\textbf{Software Announcement}

\newpage

\subsection{Abstract}\label{abstract}

The phylotypr R package implements the popular naive Bayesian
classification algorithm that is frequently used to classify 16S rRNA
and other gene sequences to taxonomic lineages. A companion data
package, phylotyprrefdata, also provides numerous versions of taxonomic
databases from the Ribosomal Database Project, SILVA, and greengenes.

\newpage

\subsection{Announcement}\label{announcement}

Since it was published in 2007, the naive Bayesian classifier has been
the most popular and performant tool for classifying 16S rRNA gene
sequences (1). The method calculates the probability distributions of
k-mers (typically 8-mers) across a reference collection and within each
genus represented in the collection. These probabilities are used within
a pseudo-bootstrapping procedure to classify unknown sequences and
assign a confidence score to that classification. The confidence scores
are used to prune the Linnaean taxonomy to the deepest possible
taxonomic level with sufficient confidence (typically 80\%). The
algorithm was been made available by the original developers as an
application coded in Java; a wrapper for the original code was available
in QIIME (2). A C++ version has been available in mothur and a Python
version in QIIME2 (3, 4). Until March 2023, users could classify
sequences with an online interface at the Ribosomal Database Project
(RDP); this interface is no longer available. The RDP developers
continue to update their code and the database through their GitHub and
Sourceforge-based repositories (5).

Considering the growing popularity of the R programming language among
microbial ecologists (6--10), I developed an R-based version of the
algorithm that is available as the phylotypr package. Users can install
phylotypr via CRAN or through the devtools package's install\_github
function. Classification consists of two steps. First, the reference
sequences and taxonomies are used to calculate kmer-based probabilities
with the build\_kmer\_database function. Users can specify their desired
kmer size when generating the database. These probabilities can be
recalculated for each R session or saved as an R data file. Their
calculation can be completed within several seconds. Second,
user-supplied sequences can be classified using the reference database
with the classify\_sequence function. Accessory filter\_taxonomy and
print\_taxonomy functions allow the user to output the results of their
classifications using a minimum confidence score threshold. A detailed
vignette is available within the phylotypr package that demonstrates how
to install the package, use its functions, and parallelize its
performance using the furrr package. The R-based execution time is
comparable to or faster than that found when using the classify.seqs
mothur function written in C++.

Many microbial ecologists have benefited from training the algorithm
using reference sequences and taxonomies curated by the RDP as well as
other providers including greengenes and SILVA (5, 11--16). For
demonstration purposes, phylotypr includes a small reference database
using version 9 of the RDP's reference. A companion data package,
phylotyprrefdata, is available on GitHub and can be installed using the
install\_github function from the devtools package. The current version
of the data package (v0.1.0) includes all publicly available versions of
the references from each of the RDP, greengenes, and SILVA references.
Because of the size of the package (150 MB), it is too large to post to
CRAN. I plan to make regular updates to the data package as new versions
of databases become available. Users can also provide their own
reference data to classify genes other than the 16S or 18S rRNA gene to
to improve classification of lineages that are poorly represented in the
references.

\subsection{Data availability}\label{data-availability}

phylotypr is available through CRAN and developmental versions are
available through the project's GitHub website
(https://github.com/mothur/phylotypr). A pkgdown version of the
documentation is hosted at (https://mothur.org/phylotypr). The
phylotyprref data package is available through the project's GitHub
website (https://github.com/mothur/phylotyprref). Both packages are
available under the MIT open source license.

\subsection{Acknowledgements}\label{acknowledgements}

phylotypr was developed as a series of videos on the Riffomonas YouTube
channel
(https://www.youtube.com/playlist?list=PLmNrK\_nkqBpIZlWa3yGEc2-wX7An2kpCL).
I am grateful to the viewers of the Riffomonas YouTube channel for their
questions, suggestions, and encouragement throughout its development.

\newpage

\subsection{References}\label{references}

\phantomsection\label{refs}
\begin{CSLReferences}{0}{1}
\bibitem[\citeproctext]{ref-Wang2007}
\CSLLeftMargin{1. }%
\CSLRightInline{\textbf{Wang Q}, \textbf{Garrity GM}, \textbf{Tiedje
JM}, \textbf{Cole JR}. 2007. Naïve bayesian classifier for rapid
assignment of rRNA sequences into the new bacterial taxonomy. Applied
and Environmental Microbiology \textbf{73}:5261--5267.
doi:\href{https://doi.org/10.1128/aem.00062-07}{10.1128/aem.00062-07}.}

\bibitem[\citeproctext]{ref-Caporaso2010}
\CSLLeftMargin{2. }%
\CSLRightInline{\textbf{Caporaso JG}, \textbf{Kuczynski J},
\textbf{Stombaugh J}, \textbf{Bittinger K}, \textbf{Bushman FD},
\textbf{Costello EK}, \textbf{Fierer N}, \textbf{Peña AG},
\textbf{Goodrich JK}, \textbf{Gordon JI}, \textbf{Huttley GA},
\textbf{Kelley ST}, \textbf{Knights D}, \textbf{Koenig JE}, \textbf{Ley
RE}, \textbf{Lozupone CA}, \textbf{McDonald D}, \textbf{Muegge BD},
\textbf{Pirrung M}, \textbf{Reeder J}, \textbf{Sevinsky JR},
\textbf{Turnbaugh PJ}, \textbf{Walters WA}, \textbf{Widmann J},
\textbf{Yatsunenko T}, \textbf{Zaneveld J}, \textbf{Knight R}. 2010.
QIIME allows analysis of high-throughput community sequencing data.
Nature Methods \textbf{7}:335--336.
doi:\href{https://doi.org/10.1038/nmeth.f.303}{10.1038/nmeth.f.303}.}

\bibitem[\citeproctext]{ref-Schloss2009}
\CSLLeftMargin{3. }%
\CSLRightInline{\textbf{Schloss PD}, \textbf{Westcott SL},
\textbf{Ryabin T}, \textbf{Hall JR}, \textbf{Hartmann M},
\textbf{Hollister EB}, \textbf{Lesniewski RA}, \textbf{Oakley BB},
\textbf{Parks DH}, \textbf{Robinson CJ}, \textbf{Sahl JW}, \textbf{Stres
B}, \textbf{Thallinger GG}, \textbf{Van Horn DJ}, \textbf{Weber CF}.
2009. Introducing mothur: Open-source, platform-independent,
community-supported software for describing and comparing microbial
communities. Applied and Environmental Microbiology
\textbf{75}:7537--7541.
doi:\href{https://doi.org/10.1128/aem.01541-09}{10.1128/aem.01541-09}.}

\bibitem[\citeproctext]{ref-Bokulich2018}
\CSLLeftMargin{4. }%
\CSLRightInline{\textbf{Bokulich NA}, \textbf{Kaehler BD},
\textbf{Rideout JR}, \textbf{Dillon M}, \textbf{Bolyen E},
\textbf{Knight R}, \textbf{Huttley GA}, \textbf{Gregory Caporaso J}.
2018. Optimizing taxonomic classification of marker-gene amplicon
sequences with QIIME 2's q2-feature-classifier plugin. Microbiome
\textbf{6}.
doi:\href{https://doi.org/10.1186/s40168-018-0470-z}{10.1186/s40168-018-0470-z}.}

\bibitem[\citeproctext]{ref-Wang2024}
\CSLLeftMargin{5. }%
\CSLRightInline{\textbf{Wang Q}, \textbf{Cole JR}. 2024. Updated RDP
taxonomy and RDP classifier for more accurate taxonomic classification.
Microbiology Resource Announcements \textbf{13}.
doi:\href{https://doi.org/10.1128/mra.01063-23}{10.1128/mra.01063-23}.}

\bibitem[\citeproctext]{ref-Liu2020}
\CSLLeftMargin{6. }%
\CSLRightInline{\textbf{Liu C}, \textbf{Cui Y}, \textbf{Li X},
\textbf{Yao M}. 2020. Microeco: An r package for data mining in
microbial community ecology. FEMS Microbiology Ecology \textbf{97}.
doi:\href{https://doi.org/10.1093/femsec/fiaa255}{10.1093/femsec/fiaa255}.}

\bibitem[\citeproctext]{ref-Buttigieg2014}
\CSLLeftMargin{7. }%
\CSLRightInline{\textbf{Buttigieg PL}, \textbf{Ramette A}. 2014. A guide
to statistical analysis in microbial ecology: A community-focused,
living review of multivariate data analyses. FEMS Microbiology Ecology
\textbf{90}:543--550.
doi:\href{https://doi.org/10.1111/1574-6941.12437}{10.1111/1574-6941.12437}.}

\bibitem[\citeproctext]{ref-McMurdie2013}
\CSLLeftMargin{8. }%
\CSLRightInline{\textbf{McMurdie PJ}, \textbf{Holmes S}. 2013. Phyloseq:
An r package for reproducible interactive analysis and graphics of
microbiome census data. PLoS ONE \textbf{8}:e61217.
doi:\href{https://doi.org/10.1371/journal.pone.0061217}{10.1371/journal.pone.0061217}.}

\bibitem[\citeproctext]{ref-Callahan2016}
\CSLLeftMargin{9. }%
\CSLRightInline{\textbf{Callahan BJ}, \textbf{McMurdie PJ},
\textbf{Rosen MJ}, \textbf{Han AW}, \textbf{Johnson AJA}, \textbf{Holmes
SP}. 2016. DADA2: High-resolution sample inference from illumina
amplicon data. Nature Methods \textbf{13}:581--583.
doi:\href{https://doi.org/10.1038/nmeth.3869}{10.1038/nmeth.3869}.}

\bibitem[\citeproctext]{ref-Dixon2003}
\CSLLeftMargin{10. }%
\CSLRightInline{\textbf{Dixon P}. 2003. {VEGAN}, a package of r
functions for community ecology. Journal of Vegetation Science
\textbf{14}:927--930.
doi:\href{https://doi.org/10.1111/j.1654-1103.2003.tb02228.x}{10.1111/j.1654-1103.2003.tb02228.x}.}

\bibitem[\citeproctext]{ref-DeSantis2006}
\CSLLeftMargin{11. }%
\CSLRightInline{\textbf{DeSantis TZ}, \textbf{Hugenholtz P},
\textbf{Larsen N}, \textbf{Rojas M}, \textbf{Brodie EL}, \textbf{Keller
K}, \textbf{Huber T}, \textbf{Dalevi D}, \textbf{Hu P}, \textbf{Andersen
GL}. 2006. Greengenes, a chimera-checked 16S rRNA gene database and
workbench compatible with ARB. Applied and Environmental Microbiology
\textbf{72}:5069--5072.
doi:\href{https://doi.org/10.1128/aem.03006-05}{10.1128/aem.03006-05}.}

\bibitem[\citeproctext]{ref-McDonald2011}
\CSLLeftMargin{12. }%
\CSLRightInline{\textbf{McDonald D}, \textbf{Price MN}, \textbf{Goodrich
J}, \textbf{Nawrocki EP}, \textbf{DeSantis TZ}, \textbf{Probst A},
\textbf{Andersen GL}, \textbf{Knight R}, \textbf{Hugenholtz P}. 2011. An
improved greengenes taxonomy with explicit ranks for ecological and
evolutionary analyses of bacteria and archaea. The ISME Journal
\textbf{6}:610--618.
doi:\href{https://doi.org/10.1038/ismej.2011.139}{10.1038/ismej.2011.139}.}

\bibitem[\citeproctext]{ref-McDonald2023}
\CSLLeftMargin{13. }%
\CSLRightInline{\textbf{McDonald D}, \textbf{Jiang Y}, \textbf{Balaban
M}, \textbf{Cantrell K}, \textbf{Zhu Q}, \textbf{Gonzalez A},
\textbf{Morton JT}, \textbf{Nicolaou G}, \textbf{Parks DH},
\textbf{Karst SM}, \textbf{Albertsen M}, \textbf{Hugenholtz P},
\textbf{DeSantis T}, \textbf{Song SJ}, \textbf{Bartko A},
\textbf{Havulinna AS}, \textbf{Jousilahti P}, \textbf{Cheng S},
\textbf{Inouye M}, \textbf{Niiranen T}, \textbf{Jain M}, \textbf{Salomaa
V}, \textbf{Lahti L}, \textbf{Mirarab S}, \textbf{Knight R}. 2023.
Greengenes2 unifies microbial data in a single reference tree. Nature
Biotechnology \textbf{42}:715--718.
doi:\href{https://doi.org/10.1038/s41587-023-01845-1}{10.1038/s41587-023-01845-1}.}

\bibitem[\citeproctext]{ref-Yilmaz2013}
\CSLLeftMargin{14. }%
\CSLRightInline{\textbf{Yilmaz P}, \textbf{Parfrey LW}, \textbf{Yarza
P}, \textbf{Gerken J}, \textbf{Pruesse E}, \textbf{Quast C},
\textbf{Schweer T}, \textbf{Peplies J}, \textbf{Ludwig W},
\textbf{Glöckner FO}. 2013. The SILVA and {``all-species living tree
project (LTP)''} taxonomic frameworks. Nucleic Acids Research
\textbf{42}:D643--D648.
doi:\href{https://doi.org/10.1093/nar/gkt1209}{10.1093/nar/gkt1209}.}

\bibitem[\citeproctext]{ref-Werner2011}
\CSLLeftMargin{15. }%
\CSLRightInline{\textbf{Werner JJ}, \textbf{Koren O}, \textbf{Hugenholtz
P}, \textbf{DeSantis TZ}, \textbf{Walters WA}, \textbf{Caporaso JG},
\textbf{Angenent LT}, \textbf{Knight R}, \textbf{Ley RE}. 2011. Impact
of training sets on classification of high-throughput bacterial 16s rRNA
gene surveys. The ISME Journal \textbf{6}:94--103.
doi:\href{https://doi.org/10.1038/ismej.2011.82}{10.1038/ismej.2011.82}.}

\bibitem[\citeproctext]{ref-Quast2012}
\CSLLeftMargin{16. }%
\CSLRightInline{\textbf{Quast C}, \textbf{Pruesse E}, \textbf{Yilmaz P},
\textbf{Gerken J}, \textbf{Schweer T}, \textbf{Yarza P}, \textbf{Peplies
J}, \textbf{Glöckner FO}. 2012. The SILVA ribosomal RNA gene database
project: Improved data processing and web-based tools. Nucleic Acids
Research \textbf{41}:D590--D596.
doi:\href{https://doi.org/10.1093/nar/gks1219}{10.1093/nar/gks1219}.}

\end{CSLReferences}




\end{document}
